internal/server/singleprocess/auth.go

```diff
@@ -57,44 +61,93 @@ var (
 	}
 )
 
+type userKey struct{}
+
+// userWithContext inserts the user value u into the context. This can
+// be extracted with userFromContext.
+func userWithContext(ctx context.Context, u *pb.User) context.Context {
+	return context.WithValue(ctx, userKey{}, u)
+}
+
+// userFromContext returns the authenticated user in the request context.
+// This will return nil if the user is not authenticated.
+func (s *service) userFromContext(ctx context.Context) *pb.User {
+	value, ok := ctx.Value(userKey{}).(*pb.User)
+	if !ok && s.superuser {
+		value = &pb.User{Id: DefaultUserId, Username: DefaultUser}
+	}
+
+	return value
+}
+
 // Authenticate implements the server.AuthChecker interface.
 //
 // This checks if the given endpoint should be allowed. This is called during a
 // gRPC request. Effects is some information about the endpoint, at present these are
 // either ["readonly"] or ["mutable"] to indicate if the endpoint will be only reading
 // data or also mutating it.
-func (s *service) Authenticate(ctx context.Context, token, endpoint string, effects []string) error {
+func (s *service) Authenticate(
+	ctx context.Context, token, endpoint string, effects []string,
+) (context.Context, error) {
 	// Ignore unauthenticated endpoints
 	if _, ok := unauthenticatedEndpoints[endpoint]; ok {
-		return nil
+		return nil, nil
 	}
 
 	if token == "" {
-		return status.Errorf(codes.Unauthenticated, "Authorization token is not supplied")
+		return nil, status.Errorf(codes.Unauthenticated, "Authorization token is not supplied")
 	}
 
 	_, body, err := s.decodeToken(token)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	// Token must be a login token to be used for auth
 	login, ok := body.Kind.(*pb.Token_Login_)
 	if !ok || login == nil {
-		return ErrInvalidToken
+		return nil, ErrInvalidToken
 	}
 
 	// If this is an entrypoint token then we can only access entrypoint APIs.
 	if login.Login.Entrypoint != nil && !strings.HasPrefix(endpoint, "Entrypoint") {
-		return status.Errorf(codes.Unauthenticated, "Unauthorized endpoint")
+		return nil, status.Errorf(codes.Unauthenticated, "Unauthorized endpoint")
 	}
 
-	// TODO When we have a user model, this is where you'll check for the user.
-	if login.Login.UserId != DefaultUser {
-		return ErrInvalidToken
+	// Look up the user that this token is for.
+LOOKUP_USER:
+	user, err := s.state.UserGet(&pb.Ref_User{
+		Ref: &pb.Ref_User_Id{
+			Id: &pb.Ref_UserId{Id: login.Login.UserId},
+		},
+	})
+	if status.Code(err) == codes.NotFound {
+		// If we are a legacy token logging into a WP 0.5+ server for the
+		// first time, we need to create the initial Waypoint user. This is
+		// purely a backwards compatibility case that we should drop at some
+		// point.
+		if !body.UnusedLogin || login.Login.UserId != DefaultUserId {
+			return nil, status.Errorf(codes.Unauthenticated, "Expired or invalid authentication token")
```

Comment:
When would an expired token make it here? I would think it would get caught on [on line 101](https://github.com/hashicorp/waypoint/pull/1752/files#diff-2155cc372482197f0003ac88874cc7fc653780a65ecea2305ab1a4b3b48d7d1eR101). Are you thinking about invite (non-login) tokens generated pre-update that haven't been redeemed?

Reply:
Also, if we merge this now, all tokens generated since #1644 will fail here because UnusedLogin is no longer set and defaults to false.

Reply:
@izaaklauer `UnusedLogin` will be set to true for old tokens (tokens issued before 0.4). 

And the expired point: it never will be expired, let me fix the error message.

Reply:
r/e `UnusedLogin`: I hit this error when I presented a token created by main, because main no longer sets `UnusedLogin`. I don't think it's a huge issue, because we didn't backport #1644, but I think it'll mean everyone running off main will have to abandon their contexts, or cut a new token using 0.4.1.

Reply:
Ah yes this is true, this is backwards incompatible with `main`, but we haven't ever really tried to maintain compatibility there, only with released versions. :)